catalog/async/catalog-async-data/src/main/java/org/codice/ddf/catalog/async/data/impl/LazyProcessResourceImpl.java

```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * Creates a {@link ProcessResource} with a {@link Supplier} used to get the a resource.
+     * The resource will not be loaded until the first time one of the following fields is accessed:
+     * inputStream
+     * mimeType
+     * name
+     *
+     * @param metacardId       schema specific part of {@link URI}, throws {@link IllegalArgumentException} if empty or null
+     * @param resourceSupplier a {@link Supplier} used to load the resource
+     * @throws IllegalArgumentException if the input provided is not valid
+     *
+     **/
+    public LazyProcessResourceImpl(String metacardId, Supplier<Resource> resourceSupplier) {
+        if (StringUtils.isBlank(metacardId)) {
+            throw new IllegalArgumentException(
+                    "ProcessResourceImpl argument \"metacardId\" may not be null or empty.");
+        }
+        if (resourceSupplier == null) {
+            throw new IllegalArgumentException(
+                    "LazyProcessResourceImpl must have a non null resource supplier");
+        }
+
+        this.isModified = false;
+        this.metacardId = metacardId;
+        this.resourceSupplier = resourceSupplier;
+        this.isResourceLoaded = false;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This will load the resource.
+     */
+    @Override
+    public InputStream getInputStream() throws IOException {
+        loadResource();
+        return inputStream;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This will load the resource.
+     */
+    @Override
+    public String getMimeType() {
+        loadResource();
+        return mimeType;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * This will load the resource.
+     */
+    @Override
+    public String getName() {
+        loadResource();
+        return name;
+    }
+
+    @Override
+    public String getQualifier() {
+        return qualifier;
+    }
+
+    @Override
+    public long getSize() {
+        return size;
+    }
+
+    @Override
+    public URI getUri() {
+        return uri;
+    }
+
+    public void setSize(long size) {
+        this.size = size;
+    }
+
+    public void setUri(URI uri) {
+        this.uri = uri;
+
+        if (uri != null && StringUtils.isNotBlank(uri.getFragment())) {
+            this.qualifier = uri.getFragment();
+        }
+    }
+
+    @Override
+    public boolean isModified() {
+        return isModified;
+    }
+
+    @Override
+    public void close() {
+        IOUtils.closeQuietly(inputStream);
+    }
+
+    public void markAsModified() {
+        isModified = true;
+    }
+
+    private void loadResource() {
+        if (!isResourceLoaded) {
+            populateProcessResource(resourceSupplier.get());
+        }
+    }
+
+    private void populateProcessResource(Resource resource) {
+        if (resource == null) {
+            LOGGER.debug(
+                    "Error loading resource for metacard id: {}. null resource was returned by supplier.",
```

Comment:
This debug statement could include a message indicating what the potential consequences of this are.
Will this cause all async processing to fail for this resource?

Reply:
Yes, this will fail async processing of this resource.

Changed to throw InaccessibleResourceException.